library/Ivoz/Provider/Domain/Service/Terminal/TerminalFactory.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace Ivoz\Provider\Domain\Service\Terminal;
+
+use Ivoz\Core\Application\Service\EntityTools;
+use Ivoz\Provider\Domain\Model\Terminal\TerminalDto;
+use Ivoz\Provider\Domain\Model\Terminal\TerminalInterface;
+use Ivoz\Provider\Domain\Model\Terminal\TerminalRepository;
+use Ivoz\Provider\Domain\Model\TerminalModel\TerminalModelRepository;
+
+class TerminalFactory
+{
+    protected $terminalRepository;
+    protected $terminalModelRepository;
+    protected $entityTools;
+
+    public function __construct(
+        TerminalRepository $terminalRepository,
+        TerminalModelRepository $terminalModelRepository,
+        EntityTools $entityTools
+    ) {
+        $this->terminalRepository = $terminalRepository;
+        $this->terminalModelRepository = $terminalModelRepository;
+        $this->entityTools = $entityTools;
+    }
+
+    /**
+     * @throws \Exception
+     */
+    public function fromMassProvisioningCsv(
+        int $companyId,
+        string $name,
+        string $password,
+        string $model,
+        string $mac
+    ): TerminalInterface {
+
+        $terminalModel = $this->terminalModelRepository->findOneByName($model);
+
+        if (!$terminalModel) {
+            throw new \DomainException(
+                'terminal model not found',
+                404
+            );
+        }
+
+        $terminal = $this->terminalRepository->findOneByCompanyAndName(
+            $companyId,
+            $name
+        );
+
+        if (is_null($terminal) && $mac) {
+            $terminal = $this->terminalRepository->findOneByMac(
+                $mac
+            );
+        }
```

Comment:
to be ultra safe here, I would add company to this find to avoid selecting a terminal of another company a create a data inconsistency.

Reply:
:+1: